controller/controller-common/src/main/java/com/pinterest/rocksplicator/controller/mysql/MySQLTaskQueue.java

```diff
@@ -17,71 +17,130 @@
- *  PRIMARY KEY (name)
- *
+ * A {@link TaskQueue} which talks to MySQL directly on all kinds of queue operations through JPA.
  */
-public class MySQLTaskQueue implements TaskQueue{
+public class MySQLTaskQueue implements TaskQueue {
+
+  class MySQLTaskQueueException extends Exception {
+    public MySQLTaskQueueException(String message) {
+      super(message);
+    }
+  }
 
+  private EntityManager entityManager;
   private static final Logger LOG = LoggerFactory.getLogger(MySQLTaskQueue.class);
-  private static final String TAG_TABLE_NAME = "tag";
-  private static final String TASK_TABLE_NAME = "task";
-  private static final String INSERT_TAG_QUERY_TEMPLATE =
-      "INSERT IGNORE INTO %s (name, locks, created_at) VALUES ('%s', %d, NOW())";
-  private static final String LOCK_CLUSTER_QUERY_TEMPLATE =
-      "UPDATE TAG SET locks=1 WHERE name='%s' AND locks=0";
-  private static final String UNLOCK_CLUSTER_QUERY_TEMPLATE =
-      "UPDATE TAG SET locks=0 WHERE name='%s' AND locks=1";
-  private final Connection connection;
 
-  MySQLTaskQueue(Connection connection) {
-    this.connection = connection;
+  public MySQLTaskQueue(EntityManager entityManager) {
+    this.entityManager = entityManager;
+  }
+
+  public boolean isClusterLocked(final String clusterName) throws MySQLTaskQueueException {
+    List<Integer> result = entityManager
+        .createNamedQuery("tag.findLock")
+        .setParameter("name",clusterName)
+        .getResultList();
+    if (result.isEmpty()) {
+      throw new MySQLTaskQueueException("Cluster doens't exist");
+    }
+    Integer locks = result.get(0);
+    return locks == 1;
   }
 
   @Override
   public boolean createCluster(final String clusterName) {
-    String insertQuery = String.format(
-        INSERT_TAG_QUERY_TEMPLATE, TAG_TABLE_NAME, clusterName, 0);
-    return JdbcUtils.executeUpdateQuery(connection, insertQuery);
+    TagEntity cluster = entityManager.find(TagEntity.class, clusterName);
+    if (cluster != null) {
+      LOG.error("Cluster {} is already existed", clusterName);
+      return false;
+    }
+    TagEntity tagEntity = new TagEntity().setName(clusterName);
+    entityManager.getTransaction().begin();
+    entityManager.persist(tagEntity);
+    entityManager.getTransaction().commit();
+    return true;
   }
 
   @Override
   public boolean lockCluster(final String clusterName) {
-    String lockQuery = String.format(LOCK_CLUSTER_QUERY_TEMPLATE, clusterName);
-    return JdbcUtils.executeUpdateQuery(connection, lockQuery);
+    boolean locked = true;
+    try {
+      locked = isClusterLocked(clusterName);
+    } catch (MySQLTaskQueueException e) {
+      LOG.error("Cluster {} hasn't been created", clusterName);
+    } finally {
+      if (locked) {
+        LOG.error("Cluster {} is already locked, cannot double lock", clusterName);
+        return false;
+      }
+    }
+    Query query = entityManager.createNamedQuery("tag.lock").setParameter("name", clusterName);
```

Comment:
What's the result of this query is the # of affected row is 0? I think we need to make sure whether the caller indeed acquire the lock or if the lock is already held by others.